cocotb/types/logic.py

```diff
@@ -0,0 +1,207 @@
+    VHDL's 9-value ``std_ulogic`` type maps to this type by treating weak values as full strength values
+    and "uninitialized" (``U``) and "don't care" (``-``) as "unknown" (``X``).
+
+    Supports common logic operations ``&``, ``|``, ``^``, and ``~``.
+    Can be converted to and from :class:`int`, :class:`str`, :class:`bool`, and :class:`cocotb.types.Bit`.
+    """
+
+    __singleton_cache__ = {}
+
+    _repr_map = {
+        # 0 and weak 0
+        False: 0,
+        0: 0,
+        "0": 0,
+        "L": 0,
+        "l": 0,
+        # 1 and weak 1
+        True: 1,
+        1: 1,
+        "1": 1,
+        "H": 1,
+        "h": 1,
+        # unknown, unassigned, and weak unknown
+        None: 2,
+        "X": 2,
+        "x": 2,
+        "U": 2,
+        "u": 2,
+        "W": 2,
+        "w": 2,
+        "-": 2,
+        # high impedance
+        "Z": 3,
+        "z": 3,
+    }
+
+    @cache
+    def __new__(cls, value=None):
+        # convert to internal representation
+        try:
+            _repr = cls._repr_map[value]
+        except KeyError:
+            raise ValueError(
+                "{!r} is not convertable to a {}".format(value, cls.__qualname__)
+            ) from None
+        # ensure only one object is made per representation
+        if _repr not in cls.__singleton_cache__:
+            obj = super().__new__(cls)
+            obj._repr = _repr
+            cls.__singleton_cache__[_repr] = obj
+        return cls.__singleton_cache__[_repr]
+
+    @cache
+    def __and__(self, other):
+        if not isinstance(other, type(self)):
+            return NotImplemented
+        return type(self)(
+            (
+                ("0", "0", "0", "0"),
+                ("0", "1", "X", "X"),
+                ("0", "X", "X", "X"),
+                ("0", "X", "X", "X"),
+            )[self._repr][other._repr]
+        )
+
+    def __rand__(self, other):
+        return self & other
+
+    @cache
+    def __or__(self, other):
+        if not isinstance(other, type(self)):
+            return NotImplemented
+        return type(self)(
+            (
+                ("0", "1", "X", "X"),
+                ("1", "1", "1", "1"),
+                ("X", "1", "X", "X"),
+                ("X", "1", "X", "X"),
+            )[self._repr][other._repr]
+        )
+
+    def __ror__(self, other):
+        return self | other
+
+    @cache
+    def __xor__(self, other):
+        if not isinstance(other, type(self)):
+            return NotImplemented
+        return type(self)(
+            (
+                ("0", "1", "X", "X"),
+                ("1", "0", "X", "X"),
+                ("X", "X", "X", "X"),
+                ("X", "X", "X", "X"),
+            )[self._repr][other._repr]
+        )
+
+    def __rxor__(self, other):
+        return self ^ other
+
+    @cache
+    def __invert__(self):
+        return type(self)(("1", "0", "X", "X")[self._repr])
+
+    __eq__ = object.__eq__
+
+    __hash__ = object.__hash__
```

Comment:
Why is this necessary?

Reply:
It isn't necessary, just wanted to be explicit. Many types override `__eq__` which causes the inherited `__hash__` to turn off.

Reply:
Sure, and that's a feature. I don't see any reason to keep these lines; by having them, we _increase_ the chance of a mistake if someone modifies `__eq__` but forgets to modify `__hash__`.